ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/results/tracing/FlatTraceGenerator.java

```diff
@@ -0,0 +1,296 @@
+    final Gas transactionIntrinsicGasCost =
+        gasCalculator.transactionIntrinsicGasCost(transactionTrace.getTransaction());
+    LOG.debug(
+        "Transaction intrinsic gas cost: {} - {}",
+        transactionIntrinsicGasCost.toLong(),
+        transactionIntrinsicGasCost.toHexString());
+    int traceFrameIndex = 0;
+    for (TraceFrame traceFrame : transactionTrace.getTraceFrames()) {
+      cumulativeGasCost.addAndGet(traceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+      if ("CALL".equals(traceFrame.getOpcode())) {
+        handleCall(
+            transactionTrace,
+            traceFrame,
+            lastContractAddress,
+            cumulativeGasCost,
+            addressVector,
+            tracesContexts,
+            traceFrameIndex);
+      } else if ("RETURN".equals(traceFrame.getOpcode()) || "STOP".equals(traceFrame.getOpcode())) {
+        handleReturn(transactionTrace, traceFrame, smartContractAddress, tracesContexts);
+      } else if ("SELFDESTRUCT".equals(traceFrame.getOpcode())) {
+        handleSelfDestruct(
+            traceFrame, lastContractAddress, cumulativeGasCost, addressVector, tracesContexts);
+      }
+      traceFrameIndex++;
+    }
+    final List<Trace> flatTraces = new ArrayList<>();
+    tracesContexts.forEach(context -> flatTraces.add(context.getBuilder().build()));
+    traceCounter.incrementAndGet();
+    return flatTraces.stream();
+  }
+
+  private static void handleCall(
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final String lastContractAddress,
+      final AtomicLong cumulativeGasCost,
+      final List<Integer> addressVector,
+      final Deque<FlatTrace.Context> tracesContexts,
+      final int traceFrameIndex) {
+    final Bytes32[] stack = traceFrame.getStack().orElseThrow();
+    final Address contractCallAddress = toAddress(stack[stack.length - 2]);
+    if (addressVector.size() <= traceFrame.getDepth()) {
+      addressVector.add(0);
+    } else {
+      addressVector.set(traceFrame.getDepth(), addressVector.get(traceFrame.getDepth()) + 1);
+    }
+    final List<Integer> traceAddress = Lists.newCopyOnWriteArrayList(addressVector);
+    final FlatTrace.Builder subTraceBuilder =
+        FlatTrace.builder().traceAddress(traceAddress).resultBuilder(Result.builder());
+    // get the next trace frame to set the gas field (gas remaining) of the sub trace
+    final TraceFrame traceFrameAfterCall =
+        transactionTrace.getTraceFrames().get(traceFrameIndex + 1);
+    final Action.Builder subTraceActionBuilder =
+        Action.createCallAction(
+            transactionTrace.getTransaction(),
+            lastContractAddress,
+            contractCallAddress,
+            traceFrame,
+            traceFrameAfterCall.getGasRemaining());
+
+    final long gasCost = cumulativeGasCost.longValue();
+    // retrieve the previous transactionTrace context
+    Optional.ofNullable(tracesContexts.peekLast())
```

Comment:
This use of optional feels a bit clunky and creates an extra optional object.  Would this work better?
```java
    if (!traceContexts.isEmpty()) {
        FlatTrace.Context previousContext = traceContexts.peekLast();
        // increment sub traces counter of previous transactionTrace
        previousContext.getBuilder().incSubTraces();
        // set gas cost of previous transactionTrace
        previousContext
            .getBuilder()
            .getResultBuilder()
            .orElse(Result.builder())
            .gasUsed(Gas.of(gasCost).toHexString());
      }

Reply:
Done.